docs/concepts/macros/sqlmesh_macros.md

```diff
@@ -1210,27 +1214,93 @@ SELECT
+#### Argument data types
+
+Most macro functions provide arguments so users can supply custom values when the function is called. The data type of the argument plays a key role in how the macro code processes its value, and providing type annotations in the macro definition ensures that the macro code receives the data type it expects. This section provides a brief description of SQLMesh macro type annotation - find additional information [below](#typed-macros).
+
+As [mentioned above](#inputs-and-outputs), argument values passed to the macro call are parsed by SQLGlot before they become available to the function code. If an argument does not have a type annotation in the macro function definition, its value will always be a SQLGlot expression in the function body. Therefore, the macro function code must operate directly on the expression (and may need to extract information from it before usage).
+
+If an argument does have a type annotation in the macro function definition, the value passed to the macro call will be coerced to that type after parsing by SQLGlot and before the values are used in the function body. Essentially, SQLMesh will extract the relevant information of the annotated data type from the expression for you (if possible).
+
+For example, this macro function determines whether an argument's value is any of the integers 1, 2, or 3:
 
 ```python linenums="1"
 from sqlmesh import macro
 
 @macro()
-def kwarg_macro(evaluator, a, b=None, c=None):
-  if b:
-    return b
-  if c:
-    return c
-  return a
+def arg_in_123(evaluator, my_arg):
+    return my_arg in [1,2,3]
 ```
 
-```sql linenums="1"
+When this macro is called, it will return `FALSE` even if an integer was passed in the call. Consider this macro call:
+
+``` sql linenums="1"
 SELECT
-  @kwarg_macro('hello', c := 'world')
-FROM table
+    @arg_in_123(1)
+```
+
+It returns `SELECT FALSE` because:
+
+1. The passed value `1` is parsed by SQLGlot into a SQLGlot expression before the function code executes and
+2. There is no matching SQLGlot expression in `[1,2,3]`
+
+However, the macro will treat the argument like a normal Python function does if we annotate `my_arg` with the integer `int` type in the function definition:
+
+```python linenums="1"
+from sqlmesh import macro
+
+@macro()
+def arg_in_123(evaluator, my_arg: int): # Type annotation `my_arg: int`
+    return my_arg in [1,2,3]
+```
+
+Now the macro call will return `SELECT TRUE` because the value is coerced to a Python integer before the function code executes and `1` is in `[1,2,3]`.
+
+Type annotations should **always** be specified for arguments with a default value because, in contrast to user-supplied values, the default is not parsed by SQLGlot before the function code executes. Type annotation ensures that both default and user-supplied argument values are of the same type in the function body.
```

Comment:
type annotations don't need to be specified with default arguments, they just need to match.

```
def foo(evaluator, column=None):
```

is fine or

```
def foo(evaluator, column = Literal.string("hello"))
```
is also fine